File: core/player/ships/enemies/enemy.js
import { clamp } from "../../../utils/math.js";
import Weapon from "../../../weapons/weapon.js";
import { spatial } from "../../../world/spatialHash.js";
import { toroidalAngle, toroidalDirection, toroidalDistance } from "../../../world/utils.js";
import { ship } from "../player.js";
import Ship from "../ship.js";
import EnemyManager from "./manager.js";

export default class EnemyShip extends Ship {
    constructor({ x, y, width, height, angle, acceleration, turnRate = 2, life, vertices, color, name, controllable, weapon, maxWeaponHeat, state = "idle", img, flameImg }) {
        super({ x, y, width, height, angle, acceleration, weapon, turnRate, life, vertices, color, name, controllable, maxWeaponHeat, img, flameImg })
        this.state = state;
        this.seekAcceleration = this.acceleration;
        this.fleeAcceleration = this.acceleration * 1.2;
    }

    destroy() {
        EnemyManager.destroy(this);
    }

    follow(obj) {
        let tangent = 0;
        if (this.state === "flee") {
            tangent = Math.PI / 2;
        } else {
            tangent = -Math.PI / 2;
        }

        const diff = toroidalDirection(obj.x, obj.y, this.x, this.y, this.angle, tangent);

        this.angle -= clamp(diff, this.turnRate);

        return diff;
    }

    // respond to close Weapon
    // alpha: weapon face ship
    // beta: weapon back ship
    // dist: dist from ship
    closeWeapon(weapon, dist, alpha) {
        if (dist > 7 ** 7) return;
        if (Math.abs(alpha - Math.PI / 2) < Math.PI / 16) {
            this.fire();
        }

        const beta = alpha + Math.PI / 2;

        this.angle -= clamp(beta, this.turnRate * 0.5);
    }

    nearByWeapon() {
        const object = spatial.query(this.x, this.y, Math.ceil(this.height / spatial.cellSize) + 1);

        for (const weapon of object) {
            if (weapon instanceof Weapon && weapon.ship !== this) {
                const dist = toroidalDistance(weapon.x, weapon.y, this.x, this.y);
                if (dist > 8 ** 8) continue;

                const alpha = toroidalDirection(weapon.x, weapon.y, this.x, this.y, this.angle);

                this.closeWeapon(weapon, dist, alpha);
            }
        }
    }

    update(t, dt, thrust = 0, turn = 0) {
        super.update(t, dt, thrust, turn);
        if (this.state !== "AI") {
            this.speed = this.speed + (this.acceleration * dt);

            this.nearByWeapon();
        }
    }

    AI({ idleDistance, fleeCondition, seekCondition, fireCondition, }) {
        if (idleDistance) {
            this.state = "idle"
            this.acceleration = this.seekAcceleration;
        } else if (fleeCondition) {
            this.state = "flee";
            this.acceleration = this.fleeAcceleration
        } else if (seekCondition) {
            this.acceleration = this.seekAcceleration;
            this.state = "seek";
        }

        if (["seek", "flee"].includes(this.state)) {
            const delta = this.follow(ship);
            if (fireCondition.others && fireCondition.func(delta)) {
                this.fire();
            }
        }
    }
}

File: void_front/game/core/player/ships/enemies/enemy.py
from __future__ import annotations
from game.core.utils.math import clamp
from game.core.weapons.weapon import Weapon
from game.core.world.spatial_hash import spatial
from game.core.world.utils import lerp, toroidalDirection, toroidalDistance, wrap
from game.core.player.ships import player
from game.core.player.ships.ship import Ship
import numpy as np

from game.core.utils.constants import FIXED_DT


class EnemyShip(Ship):
    def __init__(self, **kwargs):
        state = kwargs.pop("state", "idle")
        super().__init__(**kwargs)
        self.state = state
        self.seekAcceleration = self.acceleration
        self.fleeAcceleration = self.acceleration * 1.2

    def destroy(self):
        from game.core.player.ships.enemies.manager import EnemyManager

        EnemyManager.destroy(self)

    def follow(self, obj):
        tangent = np.pi / 2 if self.state == "flee" else -np.pi / 2
        delta = toroidalDirection(obj.x, obj.y, self.x, self.y, self.angle, tangent)
        self.angle = wrap(lerp(self.angle, self.angle - clamp(delta, self.turnRate) * FIXED_DT * self.speed_factor, 0.9), np.pi * 2)
        return delta

    def closeWeapon(self, weapon, dist, alpha):
        if dist > 7**7:
            return
        if abs(alpha - np.pi / 2) < np.pi / 16:
            self.fire()
        beta = alpha + np.pi / 2
        delta = np.atan2(np.sin(beta), np.cos(beta))
        self.angle = wrap(lerp(self.angle, self.angle - clamp(delta, self.turnRate) * FIXED_DT * self.speed_factor, 0.3), np.pi * 2)

    def nearByWeapon(self):
        object_list = spatial.query(
            self.x, self.y, np.ceil(self.height / spatial.cellSize) + 1
        )
        for weapon in object_list:
            if isinstance(weapon, Weapon) and weapon.ship is not self:
                dist = toroidalDistance(weapon.x, weapon.y, self.x, self.y)
                if dist > 8**8:
                    continue
                alpha = toroidalDirection(
                    weapon.x, weapon.y, self.x, self.y, self.angle
                )
                self.closeWeapon(weapon, dist, alpha)

    def update(self, t, dt, thrust=0, turn=0):
        super().update(t, dt, thrust, turn)
        if self.state != "AI":
            self.speed = self.speed + (self.acceleration * dt)
            self.nearByWeapon()

    def AI(
        self,
        *,
        idleDistance=None,
        fleeCondition=None,
        seekCondition=None,
        fireCondition=None,
    ):
        if idleDistance:
            self.state = "idle"
            self.acceleration = self.seekAcceleration
        elif fleeCondition:
            self.state = "flee"
            self.acceleration = self.fleeAcceleration
        elif seekCondition:
            self.acceleration = self.seekAcceleration
            self.state = "seek"

        if self.state in ["seek", "flee"]:
            delta = self.follow(player.ship)
            if (
                fireCondition
                and fireCondition.get("others")
                and fireCondition["func"](delta)
            ):
                self.fire()
